### DISCUSSION TITLE
Location encoding or embedding

### DISCUSSION THREAD

**Discussion by krsnapaudel** [WUR-AI/AgML-CY-Bench#291] [Ideas]
ML models would probably benefit from an indication of location. How do we encode location as a feature?
1. One-hot encoding of location ids?
2. https://github.com/marccoru/locationencoder: From @marccoru.
3. SirenNet: https://github.com/vsitzmann/siren/tree/master.
4. Other ideas @ronvree, @michielkallenberg, @aikepotze, @ellaampy, @lilybellesweet?

Let's tag others as and when they show interest.

**Comment by michielkallenberg** [comment]
@krsnapaudel Do I understand it correctly that this is about a means to encode location as part of the modeling, or do you want to store those encoded locations in the dataset itself? If I'm not mistaken solution 2 is an encoder that still needs to be trained (with e.g. yield as target if you like).

**Reply by krsnapaudel** [reply]
Yes it is to encode during modeling. Think of this as a feature similar to one-hot encoding of admin ids, but more concise.

**Comment by MarcCoru** [comment]
Hi all,

thanks for tagging me. Some thoughts:

a good start (easy to implement) would be using a pre-trained location encoder from SatCLIP (https://github.com/microsoft/satclip).
Weights are public and on hugging face (for instance, https://huggingface.co/microsoft/SatCLIP-ViT16-L40)

```
from huggingface_hub import hf_hub_download
from load import get_satclip
import torch

device = "cuda" if torch.cuda.is_available() else "cpu"

c = torch.randn(32, 2)  # Represents a batch of 32 locations (lon/lat)

model = get_satclip(
    hf_hub_download("microsoft/SatCLIP-ViT16-L40", "satclip-vit16-l40.ckpt"),
    device=device,
)  # Only loads location encoder by default
model.eval()
with torch.no_grad():
    emb = model(c.double().to(device)).detach().cpu()
```

otherwise, training a location encoder from scratch following 
the ICLR paper code https://github.com/marccoru/locationencoder
More info (paper + video summary) are on the project page: https://marcrusswurm.com/locationencoder/.
Here is also a video from a recent talk with more context on location encoding: https://www.youtube.com/watch?v=8zM9XzB4A1M

here is a code snippet from the readme.
```
from locationencoder import LocationEncoder

hparams = dict(
    legendre_polys=10,
    dim_hidden=64,
    num_layers=2,
    optimizer=dict(lr=1e-4, wd=1e-3),
    num_classes=1
)

# Pytorch Lightning Model
model = LocationEncoder("sphericalharmonics", "siren", hparams)

# input longitude latitude in degrees
import torch
lonlat = torch.tensor([[51.9,5.6]], dtype=torch.float32)

# forward pass
output = model(lonlat)
```
